helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -1043,23 +1043,40 @@ public static void purgeExpiredJobs(String workflow, WorkflowConfig workflowConf
    * @param dataProvider
    * @param manager
    */
-  public static void workflowGarbageCollection(WorkflowControllerDataProvider dataProvider,
+  public static void workflowGarbageCollection(final WorkflowControllerDataProvider dataProvider,
       final HelixManager manager) {
     // Garbage collections for conditions where workflow context exists but config is missing.
-    Map<String, ZNRecord> contexts = dataProvider.getContexts();
-    HelixDataAccessor accessor = manager.getHelixDataAccessor();
-    HelixPropertyStore<ZNRecord> propertyStore = manager.getHelixPropertyStore();
 
+    Set<String> existingContexts;
+    /*
+     * Here try-catch is used to avoid concurrent modification exception while doing deep copy.
+     * Map.keySet() can produce concurrent modification exception.
+     * Reason: If the map is modified while an iteration over the set is in progress, concurrent
+     * modification exception will be thrown.
+     */
+    try {
+      existingContexts = new HashSet<>(dataProvider.getContexts().keySet());
```

Comment:
@alirezazamani Is there a specific reason why we don't change the contexts map below to a concurrent hash map and make it thread safe? 
```
  private Map<String, ZNRecord> _contextMap = new HashMap<>();
``` 
What I can only see is performance concern when contexts modification is blocked by this copying. If performance is not that bad, I believe changing the contextMap to a concurrent hash map is the right fix.

Reply:
dataProvider.getContexts() is outside the scope of this class and perhaps there are performance concerns regarding this. Maybe we can consider this in next design of TF.